Add @RangeSource argument source

mureinik opened the pull request:
This PR adds a series of argument source that provide ranges of primitieve numerical values to be used with JUnit Jupiter's `@ParameterizedTest`.
The annotations are `@ByteRangeSource`, `@ShortRangeSource`, `@IntRangeSource`, `@LongRangeSource`, `@FloatRangeSource` and `@DoubleRangeSource`.

All these annotations follow the same form:

    @IntRangeSource(to, from, step, closed)

If `step` is omitted, a default of `1` is taken.
If `closed` is omitted, a default of `false` is taken.

Closes #44
Related to junit-team/junit5#581


---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

ris58h commented:
Great job! A couple of thing I would like to suggest:
1. Add `boolean closed() default false` parameter to the annotation to control inclusive/exclusive behavior of `to`  parameter. See Java's `IntStream.range` and `IntStream.rangeClosed`.
2. Support not only `int` but `long` values as well.

smoyer64 commented:
I think the right way to support ``long`` values is to use long parameters.  If the method invocation asks for an ``int`` then range check it and either throw an exception or cast it.

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/RangeSource.java:
```diff
@@ -0,0 +1,48 @@
+ *
+ * The supplied values will be provided as arguments to the annotated {@code @ParameterizedTest} method.
+ *
+ * @see ArgumentsSource
```
Once we know which version this lands in, we need a `@since` tag.

mureinik replied:
I'll add a `@since 4.0` annotation as a placeholder in the meanwhile

nicolaiparlog replied:
I unresolved this, so we keep track of it.

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,63 @@
+	/**
+	 * The beginning of the range, inclusive.
+	 */
+	private int[] values;
```
[Extensions have to be stateless](https://blog.codefx.org/design/architecture/junit-5-extension-model/#Stateless), so we can't use a field here. I recommend to compute the values in `provideArguments`.

sbrannen replied:
Actually, in contrast to an `Extension`, an `ArgumentsProvider` that implements `AnnotationConsumer` is (always?) stateful. 😉

nicolaiparlog replied:
Oh, have to catch up on that. :blush: @mureinik: Nevermind, then.

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/RangeSource.java:
```diff
@@ -0,0 +1,48 @@
+@Documented
+@ArgumentsSource(RangeSourceProvider.class)
+public @interface RangeSource {
+	/**
```
Can you add an empty line at the beginning of each class? (Weird that the formatter doesn't enforce this).

mureinik replied:
Sure thing

nicolaiparlog commented on src/test/java/org/junitpioneer/jupiter/RangeSourceTests.java:
```diff
@@ -0,0 +1,104 @@
+	 * Fill up {@code allValues} with some values.
+	 * These values will be removed as {@link #ascending(int)} and {@link #descending(int)} receive their arguments,
+	 * and at the end of the test suite, {@link #assertAllValuesSupplied()} will check that all the values were
+	 * properly provided.
```
Hah, this is very creative. :D

I'd still prefer it a little more formal, though. Can you have a look at, for example, [`DefaultLocaleTests`](https://github.com/junit-pioneer/junit-pioneer/blob/master/src/test/java/org/junitpioneer/jupiter/DefaultLocaleTests.java#L31) to see how to test extensions? It follows these steps:

* create an inner test class with the setup you want to test; these look very similar to how the user's tests will eventually look
* in _your_ tests, ask `AbstractPioneerTestEngineTests::executeTestsForClass` to execute these inner test classes
* assert proper execution

This includes misconfigurations.

nicolaiparlog commented:
Re `boolean closed() default false`, I think that's a great idea. @mureinik, you said you added this, but I didn't see it in the PR. Forgot to push? :smiley: (Happens to me all the time...)

Re `int` vs `long` (vs `double` vs `float`) - I see these options:

1. `@RangeSource` with `fromInt`, `fromLong` (`fromFloat`, `fromDouble`)
2. `@RangeSource` with `from` that accepts `long` and, if required, converts to `int` after range check (what about `float`/`double`, then?)
3. `@IntRangeSource`, `@LongRangeSource` (`@FloatRangeSource`, `@DoubleRangeSource`), each with `from`

My thoughts:

1. A little more verbose for simple cases, but I like its one-size-fits-all character. Also, it could be extended to support several parameters as long as they are of different types.
2. I liked it at first, but it doesn't seem to extend naturally to `float`/`double`, which is unfortunate.
3. Very clear cut, I like it.

What are your thoughts?

PS: the whole "covering several parameters" thingy reminds me of #68. Maybe @sormuras has something to add?

mureinik commented:
@nicolaiparlog 
WRT `closed()` - indeed, I forgot to push. How embarrassing :-( 
WRT your other feedback - first of all, thanks! I'll review and address it in the next couple of days

sbrannen commented on src/main/java/org/junitpioneer/jupiter/RangeSource.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * {@code @RangeSource} is an {@link ArgumentsSource} which provides access to a range of {@code int} values.
+ *
+ * The supplied values will be provided as arguments to the annotated {@code @ParameterizedTest} method.
```
```suggestion
 * <p>The supplied values will be provided as arguments to the annotated {@code @ParameterizedTest} method.
```

sormuras commented:
> PS: the whole "covering several parameters" thingy reminds me of #68. Maybe @sormuras has something to add?

No, this idea and PR are looking good to me. And as "regular" argument source it could be used in a `@CartesianProductTest`, soon...

nicolaiparlog commented:
@sormuras That makes sense.

@mureinik Are you still working on this? Do you want to make any changes regarding the different data types?

mureinik commented:
@nicolaiparlog sorry again for the apparent ghosting, the day job caught up to me. I'll carve out some time for this in the next couple of days.

nicolaiparlog commented:
Don't hurry because of me, I'm notoriously laggy in replying. :wink: Do it in your own time - I just wanted to know whether you still want to do this at all.

mureinik commented:
Pushed some minor fixes just so I won't lose them. Still need to fix the tests as per the given feedback and think about the type issue

mureinik commented:
Updated the patch to address feedback given here, except for the multiple types issues. I wanted to double check I was heading the right direction (especially WRT writing the tests) before I go ahead and extend the functionality.

> Re `int` vs `long` (vs `double` vs `float`) - I see these options:
> 
> 1. `@RangeSource` with `fromInt`, `fromLong` (`fromFloat`, `fromDouble`)
> 2. `@RangeSource` with `from` that accepts `long` and, if required, converts to `int` after range check (what about `float`/`double`, then?)
> 3. `@IntRangeSource`, `@LongRangeSource` (`@FloatRangeSource`, `@DoubleRangeSource`), each with `from`
> 
> My thoughts:
> 
> 1. A little more verbose for simple cases, but I like its one-size-fits-all character. Also, it could be extended to support several parameters as long as they are of different types.
> 2. I liked it at first, but it doesn't seem to extend naturally to `float`/`double`, which is unfortunate.
> 3. Very clear cut, I like it.

2 - I'm not a fan of this option. Unless I'm missing something, JUnit won't automatically handle these casts for you, so now the provider needs to inspect the method that's annotated with it, and it add a whole new dimension of complexcity

1 vs 3 - 3 is probably neater to imoplement, but it seems less intuitive as a user (at least to me). 1 seems like a more natural extension of JUnit's [`@ValueSource`](https://junit.org/junit5/docs/5.3.1/api/org/junit/jupiter/params/provider/ValueSource.html) from a user perspective. It may require some ugly validation code to make sure your `from`s and `to`s are aligned, but IMHO it's a small price to pay.

Other thoughts/opinions are more than welcome, of course.

nicolaiparlog commented:
> 1 seems like a more natural extension of JUnit's @ValueSource from a user perspective.

That's a good argument, let's go with that then.

mureinik commented:
I played around with this a bit, and unfortuntely it won't fly. Numerical annotation members are primitives, so you don't have a natural way to specify which arguments you do or don't want to use (unlike the `@ValueSource` usecade, where the members are arrays - so by default they are all `null` except the one you choose to assign values to).

Option 3 is actually quite doable, and I'll upload a patch series that attempts to do that.

nicolaiparlog commented:
Hey @mureinik, I'm alive! :blush: I just rebased your branch onto the recent master and made some formatting changes. I will now add a review. It would be awesome if, after all this time, you want to continue working on this. Either way, let me know.

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,131 @@
+
+		Method stepMethod = argumentsSourceClass.getMethod("step");
+		Number stepValue = (Number) stepMethod.invoke(argumentsSource);
+		double stepDouble = stepValue.doubleValue();
```
This conversation is lossy. `long` contains numbers that `double` can't represent ([apparently](https://stackoverflow.com/a/3793950/2525313) the smallest one is 9,007,199,254,740,993 (2^53 + 1)). Also, `float from + float step` can lead to different results than `(float) ((double) float from + (double) float step)`.

One solution would be to have a `Range` interface with methods that return `Number` and concrete implementations for each primitive type,

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,131 @@
+		double val = fromDouble;
+		while (val < toDouble && stepDouble > 0.0 || val > toDouble && stepDouble < 0.0
+				|| closedValue && val == toDouble) {
+			values.add(val);
```
Instead of putting all (of possibly very many) elements into a list, I would create the stream directly. If the `Range` interface I mentioned above implements `Iterator<Number>` (that should be straight-forward) you can turn it into a stream as follows:

```java
StreamSupport.stream(
		Spliterators.spliterator(sourceIterator, size, Spliterator.ORDERED),
		false);
```

where `size` is roughly `(to - from) / step` (with some rounding).

nicolaiparlog commented on src/test/java/org/junitpioneer/jupiter/RangeSourceProviderTests.java:
```diff
@@ -0,0 +1,285 @@
+			PreconditionViolationException e = assertThrows(PreconditionViolationException.class,
+				() -> provider.provideArguments(ec));
+			assertEquals("Expected exactly one annotaiton to provide an ArgumentSource, found 0.", e.getMessage());
+		}
```
I think there are better ways to test misconfigurations. Shouldn't it be possible to create a regular test method that is poorly annotated and use the common `executeTestsForClass` approach to verify errors?

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/ByteRangeSource.java:
```diff
@@ -0,0 +1,55 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@ArgumentsSource(RangeSourceProvider.class)
+public @interface ByteRangeSource {
```
Hm, if we want to use this source for #68, we need this source to be repeatable. No reason to implement that now, though (I guess?).

mureinik commented:
Hi @nicolaiparlog Sorry for the late replay - I'm alive too, but was traveling and didn't closely monitor this PR.
I'll try to take a look and implement your review over the weekend.

mureinik commented:
@nicolaiparlog Took a little longer to get to this than expected, but I pushed a new revision. Changes since the last revision:

- Rebased on top of the current master and made sure that each patch individually also passes spotless (i.e., essentially took you formatting improvements and applied them on the patches that broke formatting - thanks for the help BTW!)
- Refactored out the logic of creating the stream of numbers to a `Range` abstract class (in order to share functionality) and added a concrete implementation for each numeric type. I wonder if this didn't get to the point we have too many classes, and they would be better organized in sub-package, or even as inner classes. This should solve both the lossy conversion and the performance issue of filling a list with all the elements you commented on.

Open issues:
- Regarding documentation - I'm all for it, but some guidance would be appreciated. What exactly would you want to see there?
- Regarding having the annotation repeatable - I feel that #68 isn't quite fleshed out yet, so I'd rather not prematurely involve this PR in those considerations. If whoever works on it needs to make the range sources repeatable, they could as part of that PR (unless you particularly prefer it in this PR - just say no, In any case, it seems like a simple change)
- Regarding testing invalid annotations - I'll give it a look as per your suggestion. I preferred pushing a partial solution to get some feedback instead of waiting till I have something complete.

mureinik commented:
@nicolaiparlog Test refactoring done (really looks cleaner now, thanks for the pointer!), ready for your review.

nicolaiparlog commented:
Thanks for working on this, will take a look soon (ish). One random note:

> made sure that each patch individually also passes spotless

That's very diligent, but in the end, we will squash your entire contribution into a single commit, so no need to invest time into that.

mureinik commented:
Thanks @nicolaiparlog .

I noticed this PR now has merge conflicts due to some recent merge. I've pushed an updated patch series with the following changes:
- Rebased on top of the current master (the 0.4.2 release tag)
- Bumped the `junit-jupiter-params` requirement to 5.3.2 to align with the `junit-jupiter-api` requirement
- Updated the `@since` tags to 0.5, since 0.4 is clearly wrong at this point

sormuras commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,97 @@
+				argumentsSources.size() == 1,
+				() -> String.format(
+						"Expected exactly one annotation to provide an ArgumentSource, found %d.",
+						argumentsSources.size()));
```
Couldn't you make use of [findAnnotation](https://junit.org/junit5/docs/current/api/org/junit/platform/commons/support/AnnotationSupport.html#findAnnotation(java.util.Optional,java.lang.Class))

mureinik replied:
@sormuras `findAnnotation` returns the first annotation that matches the given the type. Here, we want to get all of them in order to check that conflicting range sources aren't used on the same method.

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/ByteRangeSource.java:
```diff
@@ -0,0 +1,55 @@
+	/**
+	 * Whether the range is closed (inclusive of the {@link #to()}) or not.
+	 */
+	boolean closed() default false;
```
A totally non-representative poll of three guys in Twitch chat (plus me!) showed that we all prefer `true` as a default here. It's odd in Java terms, but if I write `@RangeSource(from = 1, to = 3)`, I expect `3` to be in there. Your take on this?

mureinik replied:
Looking at other similar APIs, like `IntStream.range` or even `List.subList` or `String.substring`, the "to" argument is always exclusive. I think having "to" here inclusive by default would be confusing than having it exclusive by default.

nicolaiparlog replied:
Yeah, it's "Java conventions" vs "common sense". Let's leave it as it is and see whether users complain. We're pre-1.0, so we can break all the things! :wink:

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/ByteRange.java:
```diff
@@ -0,0 +1,27 @@
+
+	@Override
+	public Byte nextValue() {
+		return (byte) (getCurrent() + getStep());
```
When changing this to `getCurrent().byteValue() + getStep().byteValue()` (to get rid of the cast), the compiler helpfully pointed out that this may overflow. We need to discuss how to handle that. Wrap around? Or throw an exception?

At fist I thought, this can't happen because when `step`ping from `from` to `to` we'd stop early, but with `from=120`, `to=125`, and `step=10` it _could_ happen, right?

mureinik replied:
TBH, I'm not a big fan of either of these suggestions.
To an average user `ByteRangeSource(from = 120, to = 125, step = 10)` looks like a valid range source (albeit not a particularly elegant one...). We could recognize overflows by comparing the relation between the current and next value to the sign of the step (e.g., if we calculate the next value and overflow, it will be smaller than the current value, which shouldn't happen if the step is positive), and just terminate the stream. Thus, this range source will return a single value, `120`.

I'll add a patch to treat it, and then we can discuss against some code.

nicolaiparlog replied:
Just checked, the current implementation would result in a silent wrap, i.e. `120 + 10 ~> -126`. Indeed a corner case we need to discuss.

nicolaiparlog replied:
Oh, only now saw the implementation in `hasNext` - good job!

nicolaiparlog commented on src/main/java/org/junitpioneer/jupiter/Range.java:
```diff
@@ -0,0 +1,104 @@
+
+	Stream<N> asStream() {
+		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(this, Spliterator.ORDERED), false);
+	}
```
I'd prefer it this way, smaller API surface:

```java
	Stream<N> asStream() {
		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(new ValueIterator(), Spliterator.ORDERED), false);
	}

	private class ValueIterator implements Iterator<N> {

		@Override
		public boolean hasNext() {
			N nextValue = getNextValue();
			int cmp = nextValue.compareTo(to);
			return (cmp < 0 != sign < 0) || (closed && cmp == 0);
		}

		@Override
		public N next() {
			if (!hasNext()) {
				throw new NoSuchElementException();
			}
			current = getNextValue();
			return current;
		}

	}
```

mureinik replied:
Just to be clear - if you're suggesting remove `asStream` from `Range`, sure, that's not a problem, will do.

If you're suggesting separating `Range` from a new `ValueIterator` class, I'm not sure I'm following.

nicolaiparlog replied:
Yes, I was suggesting to make `ValueIterator` an inner class, so that `Range` does not implement `Iterator`. One week later, I don't really see a problem with that, though. An iterable range makes sense. Let's leave it as it is. :)

nicolaiparlog commented:
First of all, great work Allon, thanks for doing this. :+1: About your last comments:

> I wonder if this didn't get to the point we have too many classes, and they would be better organized in sub-package, or even as inner classes.

Yes, I arrived at the same conclusion. Let's put this into a package. Is `range` too generic? Should it be `parameterrange` or `argumentsrange`? Opinions welcome!

> Regarding documentation - I'm all for it, but some guidance would be appreciated. What exactly would you want to see there?

Basically an Asciidoc file in `docs` folder. But let's do that once everything else is settled.

> Regarding having the annotation repeatable - I feel that #68 isn't quite fleshed out yet, so I'd rather not prematurely involve this PR in those considerations.

Good call.

mureinik commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,97 @@
+				fromMethod.getReturnType() == toMethod.getReturnType(),
+				() -> String.format(
+						"Illegal range. Different classes of from (%s) and to (%s).",
+						fromMethod.getReturnType(), toMethod.getReturnType()));
```
It's a remnant of one of the previous iterations. The idea wat to prevent any future contributor implementing a range with conflicting `from` and `to`, but that's probably a mute point now. I'll remove it.

mureinik commented on src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java:
```diff
@@ -0,0 +1,97 @@
+		String rangeName = argumentSourceName.substring(0, argumentSourceName.length() - "Source".length());
+		Class rangeClass = Class.forName(rangeName);
+		Range<?> range = (Range) rangeClass.getConstructors()[0].newInstance(fromValue, toValue, stepValue,
+			closedValue);
```
Interesting idea.
In an ideal world, I'd like to have a `getRangeClass()` method on the actual `RangeSource`, but I'm not aware of any way to make it unmodifiable in Java. Even with a default value, that'd leave the user with an odd API where they can't really supply anything other than that default.

In the current code I tried to work around the dependency between a `Range` and a `RangeSource` with naming conventions, but I think your suggestion to make it explicit is more idiomatic to Java. I'll cook up an implementation and update this PR.

nicolaiparlog replied:
Looks good.

mureinik commented:
@nicolaiparlog @sormuras pushed a new revision with the following updates:

1. Rebased on top of master
1. Updated the copyright notices to 2020, as per 7910b5d33b6f01453ab7839c60f55cf3a466598a
1. Removed a useless check that the `to` and `from` arguments have the same type
1. Added a `@RangeClass` annotation to allow the various `RangeSource`s to map to the appropriate `Range`s without relying on naming conventions
1. Added detection for overflows, and terminated the argument stream once its detected
1. Moved `asStream` out of `Range` and in to `RangeSourceProvider`
1. Moved all the classes into the `range` subpackage. The name is a bit short, but given the context of `org.junitpioneer.jupiter` I think it's explanatory enough. This could of course be renamed if we feel there's a better name for it.

I've left all the comments unresolved so you could review and decide for yourself if the changes in this lastest set of patches indeed resolve the issues or not.

TODO: I still haven't added a documentation file. As stated by @nicolaiparlog: "But let's do that once everything else is settled."

Bukama commented on src/main/java/org/junitpioneer/jupiter/range/Range.java:
```diff
@@ -0,0 +1,100 @@
+		N nextValue = getNextValue();
+		int cmp = nextValue.compareTo(to);
+		int overflowCheck = nextValue.compareTo(current);
+		return ((cmp < 0 != sign < 0) && (overflowCheck < 0 == sign < 0)) || (closed && cmp == 0);
```
I suggest to extract these three comparisons to a boolean variable with a meaningfull each and put the variables in the return statement. At least for me it's very hard to read and @nicolaiparlog  was also struggeling on stream lately ;)

mureinik replied:
I'm open to suggestions :-)

nicolaiparlog commented:
Great, I think we're getting ready to ship this. As to documentation, have a look into the top-level `docs` folder:

* we need a `*.adoc` file explaining the feature in a few paragraphs with some code snippets (we need to be crystal clear that this builds on parameterized tests)
* we need an entry in `docs-nav.yaml`

Good work on the Javadoc, by the way! :+1:

nicolaiparlog commented:
Oh, one more thing on the package: Let's call it `params` in accordance with Jupiter's package name for parameterized tests. That also clarifies for us which future extensions belong into that package and which don't. Will you do the honors? :blush:

mureinik commented:
> Oh, one more thing on the package: Let's call it `params` in accordance with Jupiter's package name for parameterized tests. That also clarifies for us which future extensions belong into that package and which don't. Will you do the honors?

Sure, will do.

mureinik commented:
@nicolaiparlog ready for your review - renamed the `range` package to `params` as commented above, and added a first attempt to document this feature.

nicolaiparlog commented:
Thanks so much @mureinik for riding this out. Another PR that took from 2018 to 2020 to merge. :joy:

mureinik commented:
Hey, better late than never :-) Thanks for merging!
